src/parse/ris_live.go

```diff
@@ -5,45 +5,266 @@ import (
+	conn.WriteMessage(websocket.TextMessage, out1)
+
+	/* Ping message (re-send this every minute or so */
+	ping := RisMessage{"ping", nil}
+	pingstr, err := json.Marshal(ping)
+	if err != nil {
+		log.Fatal("Error marshalling ping message (!)")
+		return
+	}
+
+	for {
+		select {
+		case <-time.After(time.Duration(60) * time.Millisecond * 1000):
+			// Send an echo packet 60 seconds
+			err := conn.WriteMessage(websocket.TextMessage, pingstr)
+			if err != nil {
+				log.Println("Error during writing to websocket:", err)
+				return
+			}
+
+		case <-interrupt:
+			// We received a SIGINT; clean up
+			log.Println("Received SIGINT interrupt signal. Closing all pending connections")
+			err := conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""))
+			if err != nil {
+				log.Println("Error during closing websocket: ", err)
+				return
+			}
+
+			select {
+			case <-done:
+				log.Println("Receiver channel closed, exiting")
+			case <-time.After(time.Duration(1) * time.Second):
+				log.Println("Timeout in closing receiving channel; exiting")
+			}
+			return
+		}
+	}
+
+}
+
+//example ris message
+//{
+//	"type":"ris_message",
+//	"data": {
+//		"timestamp":1695269583.730,
+//		"peer":"217.29.66.158",
+//		"peer_asn":"24482",
+//		"id":"217.29.66.158-018ab5f0fb720005",
+//		"host":"rrc10.ripe.net",
+//		"type":"UPDATE",
+//		"path":[24482,6939,38040,23969],
+//		"community":[[24482,2],[24482,200],[24482,12000],[24482,12040],[24482,12041],[24482,20100]],
+//		"origin":"IGP",
+//		"med":0,
+//		"announcements":[{"next_hop":"217.29.66.158","prefixes":["1.1.249.0/24"]}],
+//		"withdrawals":[]
+//	}
+//}
+
+type RisAnnouncement struct {
+	NextHop  string   `json:next_hop`
+	Prefixes []string `json:prefixes`
+}
+
+type RisLiveMessageData struct {
+	Timestamp     float64           `json:timestamp`
+	Peer          string            `json:peer`
+	PeerAsn       string            `json:"peer_asn"` //,"peer_asn":"396998"
+	Id            string            `json:id`
+	Host          string            `json:host`
+	Type          string            `json:type`
+	Path          []int             `json:path`
+	Community     [][]int           `json:community`
+	Origin        string            `json:origin`
+	Med           int               `json:med`
+	Announcements []RisAnnouncement `json:announcements`
+	Withdrawals   []string          `json:withdrawals` //string of prefixes being withdrawn
+}
+
+type RisLiveMessage struct {
+	Type string             `json:"type"`
+	Data RisLiveMessageData `json:"data"`
+}
+
+func parseLiveMessage(data []byte) ([]common.BGPMessage, error) {
+	var parsedMsgs []common.BGPMessage
+	var parsedMsg common.BGPMessage
+
+	//initial format of ris live message is TYPE and DATA
+	var message RisLiveMessage
+	err := json.Unmarshal(data, &message) //decode data from JSON to a struct
+	if err != nil {
+		log.Println("Bad parse:", err)
+		log.Println("Original message:", data)
```

Comment:
should this return with an error?

Reply:
on error handling here: if you don't return out of the function after a bad parse, the rest of the function will try to process a presumably empty `message`; it looks like the outcome is only that it'll log another message later about an unhandled message type (which isn't a strictly useful error, since what actually happened was the JSON was malformed)

so you could return from here by returning either `parsedMsgs, err` (parsedMsgs should be an empty array) or `nil,err` if you can pass nil up in place of an array. The call into this function is handling the errors returned, so it'll catch anything you send back up.